pkg/controlplane/controller/legacytokentracking/controller.go

```diff
@@ -0,0 +1,211 @@
+		configMapInformer: informer,
+		configMapCache:    informer.GetIndexer(),
+		configMapSynced:   informer.HasSynced,
+		enabled:           utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LegacyServiceAccountTokenTracking),
+	}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			configMap := obj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			configMap := newObj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.freeze()
+			configMap := obj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+	})
+
+	return c
+}
+
+func (c *Controller) enqueue(configMap *corev1.ConfigMap) {
+	key, err := cache.MetaNamespaceKeyFunc(configMap)
+	if err != nil {
+		klog.Errorf("Failed to get key for ConfigMap: %v", configMap)
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *Controller) freeze() {
+	// when delete happens and the controller is enabled to create the
+	// configmap, this means there are multiple controller instances and
+	// at least one of them is not enabled to create the configmap. in
+	// such case, we should stop creating configmap in the next 30 mins.
+	// eventually, all controller instances should be either enabled or
+	// not and this condition will not be entered unless further change
+	// happens.
+	if c.enabled {
+		c.freezeUntil = time.Now().Add(freezePeriod)
+	}
+}
+
+// Run starts the controller sync loop.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting legacy_token_tracking_controller")
+	if !cache.WaitForNamedCacheSync("tokens", stopCh, c.configMapSynced) {
+		return
+	}
+
+	go c.configMapInformer.Run(stopCh)
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+	klog.Info("Ending legacy_token_tracking_controller")
+}
+
+func (c *Controller) runWorker() {
+	for c.processNext() {
+	}
+}
+
+func (c *Controller) processNext() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	if err := c.syncConfigMap(key.(string)); err != nil {
+		utilruntime.HandleError(fmt.Errorf("while syncing ConfigMap %q, err: %w", key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) syncConfigMap(key string) error {
+	_, exists, err := c.configMapCache.GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if (exists && c.enabled) || (!exists && !c.enabled) {
+		return nil
+	}
+
+	if exists && !c.enabled {
+		return c.configMapClient.ConfigMaps(metav1.NamespaceSystem).Delete(context.TODO(), configMapName, metav1.DeleteOptions{})
+	}
+
+	// at least one controller instance is deleting the configmap. no need to
+	// create the configmap.
+	if time.Now().Before(c.freezeUntil) {
+		return nil
+	}
```

Comment:
in single apiserver cluster: freeze is no-op.
- enable: apiserver should create the configmap
- disable: apiserver should delete the configmap

in multi-apiserver cluster (e.g. three): freeze is a mechanism to prevent config existing in a cluster where some apiservers are enabled while others are not. without freeze, those apiservers will fight to create/delete until all apiservers are enabled ordisabled.
- three apiservers are enabled: should create configmap.
- two enabled, one disabled: enabled apiservers should stop creation and disabled should delete.
- three apiservers are disabled: should delete configmap.

Reply:
rather than conditionally freezing based on observing delete/enabled, would it be simpler to rate-limit successful create requests? if we reach the branch where we would try to create, check `lastCreated`; if it was more recent than now()-LIMIT, skip the create and add back to our queue to reprocess at `lastCreated+LIMIT`. once we successfully create, update `lastCreated`.

Reply:
in conditionally freezing approach, it only rate limit creation when there is a recent delete. if we always rate limit creation no matter if there is a recent delete, that will be simpler for sure.

Reply:
this is done. used a rate limiter